Only mark nav menu as dialog when open

vcanales opened the pull request:
## Description

This PR addresses part of the problem described in https://github.com/WordPress/gutenberg/issues/36600, where an incorrectly labeled menu was being identified as a dialog, when in fact it was an unopened menu. This caused screen readers such as NVDA to read out the word "dialog" in places where it shouldn't have.

I would really appreciate feedback on how we can further address the problems described on the issue.

## How has this been tested?
### Navigating through a Navigation Menu which is not collapsed reads:
"Navigation landmark with N items", and then reads the text for the first link on the list.

Submenu arrows read "Button Collapsed", pressing this would open a submenu. This can be improved by more accurately describing the button with something like "Button, Collapsed Submenu".

### Navigating to a button to open the Overlay Menu reads:
"Navigation landmark, open menu button, collapsed submenu".
After pressing enter on this button, screen reader reads:

"Expanded, Dialog, List, List with N items", and then it reads the text on each item.
The menu can be navigated using the Tab key.

Commit:
only mark nav menu as dialog when open

vcanales committed:
remove redundant "menu" from toggle buttons

Redundant as screen readers such as NVDA append "Menu" to these buttons.

jasmussen commented:
Thanks for working on this. For now this PR appears to fix the issue where the navigation menu reports itself as being a dialog, even when it isn't. This happens when, for example, the navigation is set to collapse into a button to open a dialog on mobile, but show the fully expanded menu on desktop.

alexstine commented:
@vcanales Thanks for the PR. This is almost working perfectly, just a couple final suggestions.

`<button aria-expanded="false" aria-haspopup="true" aria-label="Open" class="wp-block-navigation__responsive-container-open " data-micromodal-trigger="modal-6196aa79300a0"><svg width="24" height="24" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" role="img" aria-hidden="true" focusable="false"><rect x="4" y="7.5" width="16" height="1.5"></rect><rect x="4" y="15" width="16" height="1.5"></rect></svg></button>`

This really doesn't need aria-expanded because focus is being placed in a dialog and the content outside of the dialog is non-interactive. Also, the label needs to be updated that way users know what this button does.

`<button aria-haspopup="true" aria-label="Open Menu" class="wp-block-navigation__responsive-container-open " data-micromodal-trigger="modal-6196aa79300a0"><svg width="24" height="24" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" role="img" aria-hidden="true" focusable="false"><rect x="4" y="7.5" width="16" height="1.5"></rect><rect x="4" y="15" width="16" height="1.5"></rect></svg></button>`

For the Close button:

`<button aria-label="Close" data-micromodal-close="" class="wp-block-navigation__responsive-container-close" aria-expanded="true"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" width="24" height="24" role="img" aria-hidden="true" focusable="false"><path d="M13 11.8l6.1-6.3-1-1-6.1 6.2-6.1-6.2-1 1 6.1 6.3-6.5 6.7 1 1 6.5-6.6 6.5 6.6 1-1z"></path></svg></button>`

Let's update that label and remove aria-expanded because this button simply closes the dialog. It doesn't in theory collapse anything. This is a dialog, not an accordion/tabs module/similar.

`<button aria-label="Close Menu" data-micromodal-close="" class="wp-block-navigation__responsive-container-close"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" width="24" height="24" role="img" aria-hidden="true" focusable="false"><path d="M13 11.8l6.1-6.3-1-1-6.1 6.2-6.1-6.2-1 1 6.1 6.3-6.5 6.7 1 1 6.5-6.6 6.5 6.6 1-1z"></path></svg></button>`

I am okay with keeping aria-haspopup because it can be used with dialog triggers.

> The ariaHasPopup property of the Element interface reflects the value of the aria-haspopup attribute, which indicates the availability and type of interactive popup element, such as menu or dialog, that can be triggered by an element.
https://developer.mozilla.org/en-US/docs/Web/API/Element/ariaHasPopup

Also, it seems there is an unused aria-labelledby attribute on the dialog div.

`<div class="wp-block-navigation__responsive-dialog" role="dialog" aria-modal="true" aria-labelledby="modal-6196aa79300a0-title">`

Can you add some description text such as a heading 1 right after the Close button? E.g.

`<h1 id="modal-6196aa79300a0-title">Menu</h1>`

The other option would be removing aria-labelledby and replacing it with aria-label.

`<div class="wp-block-navigation__responsive-dialog" role="dialog" aria-modal="true" aria-label="Menu">`

The whole point of implementing a label is to add context for screen readers. Otherwise, when the dialog opens, users hear "dialog" instead of "Menu dialog" which adds context.

Can these be implemented?

Thanks.

jasmussen commented:
Thank you so much for the exceptionally detailed and highly actionable feedback, Alex.

Commit:
remove aria-expanded, clarify labels

vcanales commented:
@alexstine Thanks for your thorough review! I've addressed your concerns, and will be following up with another PR tackling some issues that the Open on Click Submenus seem to have when the overlay is open (aria-expanded can be toggled, but they just remain open). 

I would appreciate one more review, and approval.

vcanales committed:
fix php linting

alexstine commented:
@vcanales It seems like aria-expanded is now gone up until you click Open navigation menu. Then it switches to `aria-expanded="true"` on click. It will then toggle to `aria-expanded="false"` after Close navigation menu is clicked. Any way to make sure that attribute doesn't get added on Open navigation menu click?

This alternate PR is working on submenus, it is possible this issue may be fixed? #36631

Thanks.

tellthemachines commented:
> Any way to make sure that attribute doesn't get added on Open navigation menu click?

`aria-expanded` is being set [here](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/navigation/view.js#L15) so it should just be a matter of deleting those lines.

> This alternate PR is working on submenus, it is possible this issue may be fixed? #36631

@alexstine if you're referring to the the issue with the open on click submenus @vcanales mentioned, that isn't fixed in #36631. 

Ideally, to fix that we should not only remove `aria-expanded` but also change the button markup to something static (a span maybe? or a heading?) when the overlay is open, because the submenus don't display as dropdowns then. On the other hand, wasn't there a discussion somewhere about rethinking all the submenus being expanded inside the overlay? If it's a large menu, it can become quite hard to read.

tellthemachines commented on packages/block-library/src/navigation/edit/responsive-wrapper.js:
```diff
@@ -36,16 +36,24 @@ export default function ResponsiveWrapper( {
+	const dialogProps = {
+		className: 'wp-block-navigation__responsive-dialog',
+		'aria-labelledby': `${ modalId }-title`,
+		...( isOpen && { role: 'dialog', 'aria-modal': true } ),
```
If we're adding these attributes dynamically only when the modal is open, do we still need the separate set of markup for when the nav is not in overlay mode? For the front end, could we toggle them in the view script instead?

tellthemachines commented on packages/block-library/src/navigation/style.scss:
```diff
@@ -522,6 +521,16 @@
+
+	&:not(.always-shown) {
+		@include break-small {
+			display: block;
```
If we want the flex layout settings to reflect correctly in the nav contents, this class has to duplicate the responsive container content styles [here](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/navigation/style.scss#L384).

tellthemachines reviewed (commented):
Thanks for working on this! Just a few comments and questions below.

Edit: when testing locally, the positioning of the URL input for nav link blocks is at the top left of the page, instead of right by the block:
<img width="1072" alt="Screen Shot 2021-11-24 at 5 42 33 pm" src="https://user-images.githubusercontent.com/8096000/143188337-7edefabe-b9a4-4b88-9d9d-8c68153db2e9.png">

This happens both in the site and the post editor. There's also a warning in the console: `Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in the componentWillUnmount method. at URLInput`. Not sure if this is related to the changes in this branch or not; rebasing trunk might help if not.

alexstine commented:
@tellthemachines I'm okay with the submenus keeping aria-expanded if they can be expanded/collapsed (open/closed). I just don't think it's needed on the main menu trigger as that opens in a dialog. Sections inside a dialog is technically a toggle, so aria-expanded is okay.

ryelle commented:
I just left a comment on https://github.com/WordPress/gutenberg/issues/36600#issuecomment-991710129, that I also noticed an issue with the dialog + VoiceOver. This PR seems to fix that issue too (thought I haven't tested any edge cases).

I noticed this broke the horizontal styling - my menu has a Page List, a Submenu, and a Custom Link, and before the PR they all line up on a single line, but after, the Page List is on its own line.

| Before (trunk) | After (PR) |
|--------|------|
| <img width="405" alt="" src="https://user-images.githubusercontent.com/541093/145685045-8f7510ed-5fca-4f1f-8fc5-8142035b44a9.png"> | <img width="293" alt="" src="https://user-images.githubusercontent.com/541093/145685050-8577465e-255e-46bc-a3cf-0c05382ebb94.png"> |

tellthemachines commented:
> I noticed this broke the horizontal styling - my menu has a Page List, a Submenu, and a Custom Link, and before the PR they all line up on a single line, but after, the Page List is on its own line.

This is likely caused by the [new container being unstyled](https://github.com/WordPress/gutenberg/pull/36617/files#r755734687).

jasmussen committed:
Try to address feedback.

jasmussen commented:
I tried to address the feedback in https://github.com/WordPress/gutenberg/pull/36617#issuecomment-992082104 and https://github.com/WordPress/gutenberg/pull/36617/files#r755734687 (thank you @tellthemachines). It isn't completely clear to me how best to test this, but I think I got it right. I would appreciate a sanity check, so we can land this one.

carolinan commented on packages/block-library/src/navigation/index.php:
```diff
@@ -526,12 +526,17 @@ function render_block_core_navigation( $attributes, $content, $block ) {
 			<div class="%5$s" style="%7$s" id="modal-%1$s">
 				<div class="wp-block-navigation__responsive-close" tabindex="-1" data-micromodal-close>
-					<div class="wp-block-navigation__responsive-dialog" role="dialog" aria-modal="true" aria-labelledby="modal-%1$s-title" >
+					<div class="wp-block-navigation__responsive-dialog" role="dialog" aria-modal="true" aria-label="Navigation Menu" >
```
This text change is missing the translation function.

tellthemachines commented:
Thanks @jasmussen ! The alignment is much better, but I think it still needs `gap` to be explicitly set for `wp-block-navigation__plain-menu`, because otherwise there doesn't seem to be any spacing at all between a page list and a navigation link.

I'm still seeing the issue with the URL input rendering at the top left of the screen, and when that happens the block toolbar also disappears (only if it's inline, if it's set to top toolbar it still shows). I'm wondering if that might be due to the two sets of markup?

jasmussen commented:
I'll take a look. One challenge I'm seeing at the moment is that in this PR I can no longer select menu items. Trunk works:
![trunk](https://user-images.githubusercontent.com/1204802/146141782-1718b4ec-d59d-4e77-aef9-21ea51780da2.gif)

In this branch, the block toolbar doesn't appear:
![cantselect](https://user-images.githubusercontent.com/1204802/146141813-4fb1b27b-9c27-41ff-966c-7ecdf81e71cb.gif)

I suspect it's the added wrapper that's intercepting a pointer event. I'll push a rebase, then see what I can do.

tellthemachines commented:
Yeah, I'd like to try adding the dialog dynamically in the front end because if it works, we won't need the two sets of markup. Hopefully I have some time to look into it tomorrow.

jasmussen commented:
The selection issue is a bit gnarly, and I don't think I can fix that in CSS only:
![click](https://user-images.githubusercontent.com/1204802/146151272-b9e427a5-fde5-48e3-9093-7ff5d37273c5.gif)

It does seem like the best solution might be a different approach to this one, which minimizes the markup changes in the editor. I'd love to help in any way I can here.

tellthemachines commented:
Ok, I created #37434 as an alternative without markup changes (apart from adding/removing attributes, of course). Feedback and testing appreciated! Ideally we should get this in before Beta 4 early next week.

Mamaduka commented:
Should we remove this from WP 5.9 Must-Haves project now that #37434 is merged?

tellthemachines commented:
Yup, this can be closed now. Thanks everyone for your collaboration!
